### PR TITLE
Conform to spec by reading as pyarrow.Table not pyarrow.RecordBatch

### PR DESCRIPTION
**Pull request by johnkerl**
This is the first in a group of three related PRs:

* This PR: conform to https://github.com/single-cell-data/SOMA/blob/main/abstract_specification.md by having `read` return`pyarrow.Table` not `pyarrow.RecordBatch` (as in an outdated version of that spec)
* #359: Forward-port https://github.com/single-cell-data/TileDB-SOMA/pull/273 from `main-old` which will truly have ASCII columns, obviating the need for our `util_arrow.ascii_to_unicode_pyarrow_readback`
* #360: Drop in C++ acceleration-library support for `read` methods, which will go in cleanly now
  * The C++ code returns `pyarrow.Table` and with the first PR our unit tests will be ready to go
  * When the C++ code reads Unicode cells it returns them as `pyarrow.LargeBinaryArray` (needing decode) but when we are properly writing ASCII cells via the Python write path then the C++ code will read ASCII cells and return them as strings (no longer needing decoding)

These three changesets could be done in a single PR, but that would be unmerciful to the reviewers.

### CODE REVIEW

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_dataframe.py:223]: This will go away entirely in another PR in this set of three, as noted in this PR's description field

File: apis/python/src/tiledbsoma/soma_indexed_dataframe.py
Comment (johnkerl) [inline apis/python/src/tiledbsoma/soma_indexed_dataframe.py:262]: ditto

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/util_arrow.py:143]: This routine is obsolete  and can be removed.  You can directly use pyarrow.concat_tables -- it accepts an iterator, so there is no benefit to creating a list to then hand to pyarrow.concat_tables.

In fact, there is a downside - this code requires that all tables be fully materialized prior to concatenation.

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/util_arrow.py:132]: Just confirming this is known to work with chunked arrays (which can happen for very large query results, ie, column size > 2GB)

File: apis/python/src/tiledbsoma/soma_sparse_nd_array.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/soma_sparse_nd_array.py:264]: Just a heads up: this hard-wired 2D assumption is a bug (fixed in `main`).  Need to make sure this is merged correctly with main. 

I'd like to re-review post-merge.

**Review by bkmartinjr — DISMISSED** [review]
Comments:
* SOMA*DataFrame - need unit tests for the full read/write API (eg, there are no Pandas unit tests for SOMADataFrame.read_as_pandas, etc).
* remove util_arrow.concat_tables and use pyarrow.concat_tables directly.  They are sematically equivalent, but the pyarrow implementation is potentially more efficient for very large queries, as it does not require the pre-alloc of all tables before concatenating.
Also, I'd like to re-review after main has been merged/reconcilied with this branch, as there are a lot of pending changes to the NdArray code.

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/util_arrow.py:132]: @bkmartinjr to be clear the _column_ is not the `pa.LargeBinaryArray` here but rather the _cell value_

File: apis/python/src/tiledbsoma/soma_sparse_nd_array.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/soma_sparse_nd_array.py:264]: @bkmartinjr i indeed have recently merged PRs to rebase onto! :)

**Comment by johnkerl** [comment]
Converting to draft while I rebase -- lots of great activity to merge in from the last day or two of commits :)

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/util_arrow.py:132]: also reminding this will go away entirely on #359

Comment (johnkerl) [reply apis/python/src/tiledbsoma/util_arrow.py:132]: (that's how Unicode encoded as bytes reads back in Arrow unless we actively decode it -- hence #359 which will remove that need)

**Comment by johnkerl** [comment]
rebased

**Comment by johnkerl** [comment]
> SOMA*DataFrame - need unit tests for the full read/write API (eg, there are no Pandas unit tests for SOMADataFrame.read_as_pandas, etc).

@bkmartinjr 100% agreed but this PR (Arrow `RecordBatch` -> `Table`) isn't the right place for it.

**Comment by johnkerl** [comment]
@bkmartinjr ready for round two :)

**Review by thetorpedodog — COMMENTED** [review]
This is pretty straightforward, and everything looks good style-wise, so I leave it to others to judge the content of the change. LGTM but not going to hit the "approve" button so that it doesn't spuriously show up as submittable. Please bump me if you want any more feedback!

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/soma_dataframe.py:172]: why not type the `ids` (aka coordinates) parameter to something more specific?  See for example SOMADenseNdArray.

I think you can literally use the `SOMADenseCoordinates` type (in types.py) as this is a dense dataframe, which only takes a scalar or slice.

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/soma_dataframe.py:172]: Also, we need to decide on whether or not `ids` is optional.  The behavior is inconsistent between DataFrame and NdArray - in the former, None is "all", in the latter, None is not accepted (only the explicit `slice(None)` is accepted).

CC: @thetorpedodog - if you want to propose a standard convention, we can pull it through the entire package.

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/soma_dataframe.py:227]: I'm not sure I understand this last `TODO` comment - we aren't dropping columns, but rather transcoding them to unicode.  Maybe clarify the comment?

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/soma_dataframe.py:246]: is this comment still valid?  The PyArrow documentation describes concat_tables, so I'm not sure we need to recapitulate it here.

Or perhaps reword to describe the intent, eg, "Concatenate all partial read results into a single Table"?

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/soma_dataframe.py:244]: `ids` parameter needs a type.  See comment on `read`

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/soma_dataframe.py:172]: If we want to make it not an Optional value, we can easily preserve the "get everything if unspecified" behavior by doing

`ids: Union[Sequence[int], slice] = slice(None)`

I like this solution because it ends up with an explicit default with already-specified semantics.

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/soma_dataframe.py:227]: yeah it's old & i'll remove it

not sure how it's showing up as "new", must be a copypasta

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/util_arrow.py:133]: I think the preferred pyarrow approach is to use the pyarrow.types helper functions rather than `isinstance`.  in this case, I believe that would be `pyarrow.types.is_large_binary`

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/util_arrow.py:132]: I believe you misunderstood my question.  A column in a table is ChunkedArray.  If there is more than one chunk in the table column, is it correctly handled?  

I believe I have convinced myself it is fine, as `pa.array` on line 134 will concatenate all of the chunks into a single `Array`.

**Review by bkmartinjr — COMMENTED** [review]
comments:
* print's in the test_soma_indexed_dataframe.py that should be removed (eg, line 108-113)
* the two dataframe objects would benefit from having some tests that check if their write methods raise the expected error when the `value` param is unexpected (eg, not a Table, or does not match the DataFrame schema).

**Review by bkmartinjr — APPROVED** [review]
The conversion from RecordBatch -> Table looks fine.  Other refinement suggestions left inline which I believe would help the code.  I leave it to you if you want to fix in this PR, in another PR, or file bugs.

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/soma_dataframe.py:172]: @bkmartinjr @thetorpedodog thanks!

Optionality of `ids` is orthogonal to this PR which is solely `pa.RecordBatch` -> `pa.Table`, and has been open long enough, and rebased enough times -- I created https://github.com/single-cell-data/TileDB-SOMA/issues/374 to track that

File: apis/python/src/tiledbsoma/util_arrow.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/util_arrow.py:133]: @bkmartinjr will do, but recall this is going away on the very next PR in this stack (#359)

File: apis/python/src/tiledbsoma/soma_dataframe.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/soma_dataframe.py:244]: https://github.com/single-cell-data/TileDB-SOMA/issues/374 also

**Review by aaronwolen — COMMENTED** [review]
🚀

**Pull request merged**
